---
title: "Actividad 4"
output:
  github_document:
    toc: true
  pdf_document:
    toc: true
---

\newpage

Análisis de biología computacional `  BT1013.525`

```
Bryan Manuel De la O Perea  A01246337
Andrés Sarellano Acevedo    A01245418
Maximiliano Villegas García A01635825
Víctor Manuel Puga Ruiz     A01568636
```

# Procedure

These are the steps we followed in order to identify the phylogenetic tree of these
variants of the SARS-CoV-2 virus:

- Collected all the sequences (5 variants + the reference genome).
- Merged all the `.fastas` in a single file.
- Used the Clustal Omega tool (<https://www.ebi.ac.uk/Tools/msa/clustalo/>) to align the sequences.

> Our results can be found in <https://www.ebi.ac.uk/Tools/services/web/toolresult.ebi?jobId=clustalo-I20210425-021750-0190-3504822-p2m> and <https://www.ebi.ac.uk/Tools/services/web/toolresult.ebi?jobId=clustalo-I20210423-005428-0562-5987948-p1m> (will only be available for a limited time).

- Used the aligned sequences to calculate the "Pairwise Distances from DNA Sequences".
- Used these results to build the tree with "Neighbor-Joining Tree Estimation".

<br/>

- **NOTE:** We repeated the process with the sequences provided by the activity, 
and  another analysis for the sequences we gathered in the previous activity.

## Dependencies

```{r}
suppressMessages(library(seqinr))
suppressMessages(library(Biostrings))
suppressMessages(library(ape))
suppressMessages(library(ggplot2))
suppressMessages(library(ggtree))
```

```{r, setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(fig.align = "right", fig.width = 9)
```


\newpage

# Analysis 1

```{r}
accessions <- c(
  "JX869059", "AY508724", "MN908947", "AY390556", 
  "AY278489", "MN985325", "AY485277", "MT292571"
)

read.file <- function(loc) {
  my_data <- readChar(loc, file.info(loc)$size)
  cat(my_data)
}

plot.colors = c(
    "-" = "#000000",
    "a" = "#106BFF",
    "t" = "#FC2B2D",
    "g" = "#30D33B",
    "c" = "#FECF0F",
    "n" = "#FD8D0E"
)

```

## Before alignment
```{r}
all.seq = read.dna("./virus/MERGED.fasta", format = "fasta")
all.seq
```
\newpage

## Results of sequence alignment (showing only first 233)
```{r}
read.file("./virus/aligned-first-places.txt")
```
\newpage

## After alignment

```{r}
clust <- read.alignment(
  "./virus/clustalo-I20210425-021750-0190-3504822-p2m.clustal_num",
  format = "clustal", forceToLower = TRUE,
)
dna <- as.DNAbin(clust)
dna
```
\newpage

## Alignment of the first 1,000 bases
```{r, fig.height = 5, fig.width = 10 }
image.DNAbin(dna[1:1000])
```

## Complete alignment
```{r, fig.height = 5, fig.width = 10 }
image.DNAbin(dna)
```
\newpage

## Difference matrix (between sequences)

```{r, fig.height = 7}
D <- dist.dna(dna)
D.mat <- as.matrix(D)

heatmap(D.mat, margins = c(10, 10))
```
\newpage

```{r, fig.height = 4.5}
tree <- nj(D)
plot.phylo(tree, main = "Phylogenetic Tree Results")
```

```{r, fig.height = 2.5}
plot.phylo(
  unroot(tree), 
  type = "unrooted", 
  no.margin = TRUE,
  use.edge.length = FALSE ## not to scale in order to avoid overlaps
)
```

\newpage

# Using `ggplot2`

```{r, fig.height = 2.5}
ggtree(tree) + ggtitle("Phylogenetic Tree Results Using ggplot") +
  geom_tippoint() + geom_tiplab() + xlim(0, 0.8)
```
```{r}
msaplot(p=ggtree(tree), fasta = dna, color=plot.colors, offset=0.3) +
  ggtitle("Tree + Sequence Alignment") +
  geom_tiplab()
```
\newpage

# Analysis 2: Different variants of coronavirus

```{r}
accessions <- c(
  "MN908947.3" = "Reference genome",
  "MW913791.1" = "B.1.1.7",
  "FR990267.1" = "B.1.351",
  "FR990291.1" = "P.1",
  "MW453109.1" = "B.1.427",
  "MW778462.1" = "B.1.429" 
)
```

## Before alignment
```{r}
all.seq = read.dna(
  "../SARS-CoV-2 Sequences (Sample)/SARS-CoV-2 MERGED.fasta", format = "fasta"
)
all.seq
```
\newpage

## Results of sequence alignment (showing only first 360)
```{r}
read.file("../SARS-CoV-2 Sequences (Sample)/aligned-first-360-places.txt")
```
\newpage

## After alignment

```{r}
clust <- read.alignment(
  "../SARS-CoV-2 Sequences (Sample)/clustalo-I20210423-005428-0562-5987948-p1m.clustal_num", 
  format = "clustal", forceToLower = TRUE,
)
dna <- as.DNAbin(clust)
dna
```
\newpage

## Alignment of the first 1,000 bases
```{r, fig.height = 5, fig.width = 10 }
image.DNAbin(dna[1:1000])
```

## Complete alignment
```{r, fig.height = 5, fig.width = 10 }
image.DNAbin(dna)
```

\newpage
## Difference matrix (between sequences)

```{r, fig.height = 7}
D <- dist.dna(dna, model = "TN93")
D.mat <- as.matrix(D)

rows <- sapply(row.names(D.mat), function(i) { paste(accessions[[i]], "  (", i, ")", sep = "") })
cols <- sapply(colnames(D.mat),  function(i) { paste(accessions[[i]], "  (", i, ")", sep = "") })

heatmap(D.mat, margins = c(20, 20), labRow = rows, labCol = cols)
```

\newpage

```{r, fig.height = 4.5}
tree <- nj(D)
labs <- sapply(tree$tip.label, function(i) { paste(accessions[[i]], "  (", i, ")", sep = "") })
tree$tip.label <- labs;

plot.phylo(tree, main = "SARS-CoV-2 Phylogenetic Tree")
```

```{r, fig.height = 2.5}
plot.phylo(unroot(tree), type = "unrooted", no.margin = TRUE)
```

\newpage

# Using `ggplot2`

```{r, fig.height = 2.5, fig.width=9}
ggtree(tree) + ggtitle("SARS-CoV-2 Phylogenetic Tree Using ggplot") +
  geom_tippoint() + geom_tiplab() + xlim(0, 0.0012)
```
```{r, , fig.width=9}
tree$tip.label <- names(tree$tip.label)
msaplot(p=ggtree(tree), fasta = dna, color=plot.colors, offset = 0.0005) +
  ggtitle("Tree + Sequence Alignment") +
  geom_tiplab()
```
\newpage

# Key learnings

- In the first analysis, some samples of SARS-CoV-2 resemble each other a lot.
They were probably from an outbreak in the same region in similar times.
  - This is illustrated by the branches in the tree from the same node that have
  similar lengths.
  
- In the second analysis, it is shown how the different classified variants of 
the virus are related to each other:
  - The Reference Genome from Wu Han is _most likely_ their ancestor.
  - The other variants are _newer_, as they are farther in the tree.
  - Our research had previously stated that the `B.1.427` and `B.1.429` variants 
    were closely related. 

# References

- _Module 24: An Intro to Phylogenetic Tree Construction in R_. 
Retrieved on April 22, 2021, 
from <https://fuzzyatelin.github.io/bioanth-stats/module-24/module-24.html>

- _Pairwise Distances from DNA Sequences_. 
Retrieved on April 22, 2021, from
<https://www.rdocumentation.org/packages/ape/versions/5.4-1/topics/dist.dna>

- _Tajimas D for sequences of different length_. 
Retrieved on April 22, 2021, from
<https://stackoverflow.com/questions/33301632/tajimas-d-for-sequences-of-different-length>

- _Visualizing and Annotating Phylogenetic Trees with R+ggtree_. 
Retrieved on April 23, 2021, from
<https://4va.github.io/biodatasci/r-ggtree.html>



