```{r}
# navigate to your working directory (varies per user)
setwd("/cloud/project/Activities")
options(connectionObserver = NULL)

install <- function(lib) {
  if (!requireNamespace(lib, quietly = TRUE)) {
    install.packages(lib)
  }
}

install("ape")
#install("phytools")
install("seqinr")

if (FALSE) {
install("BiocManager")
BiocManager::install("Biostrings")
library(Biostrings)
  
}


#install("phangorn")
#install("geiger")

library(ape)
#library(phytools)
library(seqinr)
#library(phangorn)
#library(geiger)
```

```{r}
reference <- read.fasta(
  "SARS-CoV-2 Sequences (Sample)/SARS-CoV-2 (Reference).fasta",
  as.string = TRUE, seqonly = TRUE
)[[1]]

B.1.1.7   <- read.fasta(
  "SARS-CoV-2 Sequences (Sample)/SARS-CoV-2 (B.1.1.7) Sample.fasta",
  as.string = TRUE, seqonly = TRUE
)[[1]]

```
```{r}

library(seqinr)
if (TRUE) {
install("BiocManager")
BiocManager::install("Biostrings")
library(Biostrings)
}
library(ape)
install("textmineR")
library(textmineR)

d <- c("SARS-CoV-2 Sequences (Sample)/SARS-CoV-2 (Reference).fasta","SARS-CoV-2 Sequences (Sample)/SARS-CoV-2 (B.1.1.7) Sample.fasta")

nfiles = length(d)
seqdat <- vector("list", nfiles)

for(i in 1:nfiles){
seqdat[[i]] <- read.fasta(file=d[i])
}

label <- sapply(1:nfiles, function(k) unlist(strsplit(d[k], "[.]"))[1])
names(seqdat) <- label

seqdat_join <- lapply(seqdat, function(k) paste(toupper(unlist(k)),collapse="") )



#enumerate all 5-mers
dna <- c("A","C","G","T")
kmer5 <- expand.grid(dna, dna, dna, dna, dna)
kmer5 <- apply(kmer5, 1, function(k) paste(k, collapse=""))

#function for counting all possible kmers (k=5) given a single dna string
kmercount <- function(data){
  sapply(1:length(kmer5), function(k)
  length(unlist(gregexpr2(kmer5[k], data)))
  )}

#vector of counts for all possible kmers (k=5) for all viral sequences
kmer_features <- lapply(seqdat_join, function(k) kmercount(k))


#Collect k-mer counts into a data frame
M <- do.call(rbind, kmer_features)
      


#taxonomic labels
taxonomy <- data.frame(rownames(M), 
c("Filoviridae", "Flaviviridae"))
colnames(taxonomy) <- c("Virus","Family")

#Simplify virus species names
virusnames <- sapply(1:nrow(taxonomy), function(k){
chop <- unlist(strsplit(as.character(taxonomy$Virus)[k],"_")) 
chop[length(chop)]}
)

rownames(M) <- virusnames

tipcolor <- c("red","blue","darkviolet")[unclass(taxonomy$Family)]
      
#The correct input for CalcJSDivergence is the (unnormalised) count vector
JSdist <- CalcJSDivergence(M)

plot.phylo(bionj(JSdist), type="unrooted", cex=0.8, tip.color=tipcolor,
rotate.tree=95)
      
      
```

